### ISSUE TITLE
Every weapon on whitelist jams constantly

### ISSUE THREAD

**Issue by Will-Nichols** [diwako/ACE-punish-unknown-weapon#1]
Below is what I have for the whitelist that I am testing in a mission. I entered the names of the weapons exactly as they are in the game and they still jam about every 2-5 rounds.

`// Punish unknown weapon
force force diw_unknownwp_add_weapons = "vme_pla_qcw05,vme_pla_qcw85vme_pla_LR4,LMG_03_F,SMG_02_F,rpa_weap_m3a1,SMG_05_F,arifle_TRG21_F.arifle_TRG20_F,arifle_TRG21__GL_F,srifle_DMR_06_olive_F,srifle_DMR_06_camo_F";
force force diw_unknownwp_briefing = true;
force force diw_unknownwp_cooldown = 60;
force force diw_unknownwp_dispersion_add = 25;
force force diw_unknownwp_enable = true;
force force diw_unknownwp_jam_explosion = 2.5;
force force diw_unknownwp_jamchance_add = 9.69248;
force force diw_unknownwp_propagation = false;
force force diw_unknownwp_reload_failure = 1.3369;`

**Comment by diwako** [comment]
Thanks for the report, I am going to investigate this. Do you have by any chance a sample mission?

I also notice you have dot instead comma in your additional whitelist. `arifle_TRG21_F.arifle_TRG20_F` and there is a class name with double underscore `arifle_TRG21__GL_F`.
Just for my understanding, you are filling the additional whitelist with guns that players will not have at mission start, correct?

**Comment by Will-Nichols** [comment]
Yes; I am attempting to get that done. I didn't see the "." in there so that may be the reason the whitelist isn't working. I also don't have a test mission as I am working on this right in the beginning of a large mission. 

Does the script automatically give guys the ability to use weapons that they are scripted to load with? If so we have some issues with SMA weapons not working properly then as they are jamming every 4-6 rounds.

**Comment by diwako** [comment]
Yes, the cooldown timer in the cba settings is the time the server will wait and then will add all primary weapons of each players as a known weapon. After that time has passed every JIP player will send the server their current main gun in the case it is not in the global whitelist.
I have not used SMA weapons so far, might be a incompatibility? Not sure.

**Comment by Will-Nichols** [comment]
You are more than welcome to download the mission file and the Arma 3 preset from our GitHub page for the mission I am working on at this time. If you want to take a look at the whole thing.

[https://github.com/Task-Force-Red/Operation-Tanoa-Takedown](url)

**Comment by diwako** [comment]
Thanks I will check it out once I find the time, right now I am busy with mission polishing. I plan on rewriting parts of this script and putting it into a mod some time next week, so I will try to find this issue and fix it as well.

**Comment by diwako** [comment]
@Will-Nichols I have taken a quick look at your mission frameworks cba settings,  I don't know why I did not see it before as you actually posted it in the issues head.
You are setting `diw_unknownwp_propagation` to false, this means only the local weapon white list will be used. That however is not a CBA setting and needs to be executed in initPlayerlocal or in a postInit function.
The variable name for the array is `diw_unknownwp_local_weapons` and must be set on every client. This whole set up is an advanced mission maker thing as it is actually meant for TVTs rather than coop missions.
So if you are in a coop mission you can just enable the propagation (unless you have some special mission theme going or so).

In short:
Disable propagation means, no global whitelist, even if you add weapons to `diw_unknownwp_add_weapons`. This only adds things to the global whitelist which will not be used then.
With Propagation disabled fill `diw_unknownwp_local_weapons`  locally on a client. Also please check the information about this variable in readme.md for additional information.

Tell me if that fixes your problem.

**Comment by Will-Nichols** [comment]
Will do...going to be a little bit before I can take a swing at it.

**Comment by diwako** [comment]
That is fine keep me updated, i leave this issue open till then.

**Comment by Will-Nichols** [comment]
I have tried it as you suggested and either I am not doing something correctly or it isn't working. I have to pull it out of my mission so I can make the mission function. I can always add the jamming functionality back in IF I can get it working correctly.

The premise of my mission is to have the SF team in the mission as the ones who supply the weapons and teach the none SF players about weapons. Not the insurgents picking up and using all the weapons they can get their hands on and basically not needing to communicate with the SF team.

**Comment by Will-Nichols** [comment]
Giving it another go right now.

**Comment by diwako** [comment]
On that premise, it exceeds the scope of this script. If you want something like teaching non SF players the weapons in order for those to become "known" weapons you need to work around a bit.
Disable propagation and you need to fill the variable `diw_unknownwp_local_weapons` for each player separately. The SF players will have all weapons while the non SF players will not.

Once you consider them trained you need do following on one machine to add it to the global whitelist and clear existing disadvantages.

```sqf
diw_unknownwp_weapon_whitelist pushBackUnique toUpper(_weapon);
[_weapon,{ace_overheating_cacheWeaponData setVariable [_this,nil]}] remoteExec ["call"];
publicVariable "diw_unknownwp_weapon_whitelist";
```

**Comment by Will-Nichols** [comment]
Basically all I want to do is to have the non-SF players never be able to pick up all the weapons from the enemy they kill and force them to be dependent upon another group giving them weapons and or cash money to purchase limited black market weapons. So with that said as long as I put the array in my cba_settings.sqf file (which every player has due to it being bundled into the pbo of the mission file) and have all the classnames CAPITALIZED while enabling propagation I should be good? Or do I disable propagation?

Just to be clear though...If I want to add weapons to the whitelist for the start of the mission I put them in the portion of the script as the following:
`diw_unknownwp_add_weapons = "VME_PLA_QCW05,VME_PLA_QCW85,VME_PLA_LR4,LMG_03_F,SMG_02_F,RPA_WEAP_M3A1,SMG_05_F,ARIFLE_TRG21_F,ARIFLE_TRG20_F,ARIFLE_TRG21__GL_F,SRIFLE_DMR_06_OLIVE_F,SRIFLE_DMR_06_CAME_F";`

I really appreciate all your help with this. As far as being a mission builder I do consider myself pretty advanced but sometimes scripting in SQF just blows my mind up. Its been a real long time since I wrote code and even though I can read and understand the premise of it; I have a hard time putting my own together to function properly.

**Comment by diwako** [comment]
As this issues is now really a bug I am close this. We are going to talk about this on steam chat as this scope is different than the actual intent of the script.
